plugins/spectate.cpp

```diff
@@ -0,0 +1,157 @@
+#include <modules/EventManager.h>
+#include <modules/Job.h>
+#include <modules/Units.h>
+#include <df/job.h>
+#include <df/unit.h>
+#include <df/historical_figure.h>
+#include <df/global_objects.h>
+#include <df/world.h>
+#include <df/ui.h>
+
+#include <map>
+#include <set>
+#include <random>
+
+std::map<uint16_t,uint16_t> freq;
+std::set<int32_t> job_tracker;
+std::default_random_engine RNG;
+
+//#include "df/world.h"
+
+using namespace DFHack;
+using namespace df::enums;
+
+DFHACK_PLUGIN("spectate");
+DFHACK_PLUGIN_IS_ENABLED(enabled);
+bool following_dwarf = false;
+void* job_watched = nullptr;
+int32_t timestamp = -1;
+REQUIRE_GLOBAL(world);
+REQUIRE_GLOBAL(ui);
+
+command_result spectate (color_ostream &out, std::vector <std::string> & parameters);
+
+DFhackCExport command_result plugin_init (color_ostream &out, std::vector <PluginCommand> &commands) {
+    commands.push_back(PluginCommand("spectate",
+                                     "Automated spectator mode.",
+                                     spectate,
+                                     false,
+                                     ""
+                                     " spectate\n"
+                                     "    toggles spectator mode\n"
+                                     "\n"));
+    return CR_OK;
+}
+
+DFhackCExport command_result plugin_shutdown (color_ostream &out) {
+    return CR_OK;
+}
+
+void onTick(color_ostream& out, void* tick);
+void onJobStart(color_ostream &out, void* job);
+void onJobCompletion(color_ostream &out, void* job);
+
+DFhackCExport command_result plugin_enable(color_ostream &out, bool enable) {
+    namespace EM = EventManager;
+    if (enable && !enabled) {
+        out.print("Spectate mode enabled!\n");
+        using namespace EM::EventType;
+        EM::EventHandler ticking(onTick, 15);
+        EM::EventHandler start(onJobStart, 0);
+        EM::EventHandler complete(onJobCompletion, 0);
+        EM::registerListener(EventType::TICK, ticking, plugin_self);
+        EM::registerListener(EventType::JOB_STARTED, start, plugin_self);
+        EM::registerListener(EventType::JOB_COMPLETED, complete, plugin_self);
+        //out.print("spectator mode enabled!\n");
+    } else if (!enable && enabled) {
+        out.print("Spectate mode disabled!\n");
+        EM::unregisterAll(plugin_self);
+        job_tracker.clear();
+        freq.clear();
+        //out.print("spectator mode disabled!\n");
+    } else {
+        return CR_FAILURE;
+    }
+    enabled = enable;
+    return CR_OK;
+}
+
+command_result spectate (color_ostream &out, std::vector <std::string> & parameters) {
+    return plugin_enable(out, !enabled);
+}
+
+void onTick(color_ostream& out, void* ptr){
+    int32_t tick = df::global::world->frame_counter;
+    if(!following_dwarf || (job_watched == nullptr && (tick - timestamp) > 50)) {
+        std::vector<df::unit*> dwarves;
+        for (size_t i = 0; i < df::global::world->units.active.size(); i++) {
```

Comment:
Since you don't actually use `i` except to index into the vector that you're iterating through, foreach syntax would be cleaner:
`for (auto unit : world->units.active)`

`std::copy_if` might be even cleaner, but only use it if you're comfortable with it.

Reply:
copy_if looks interesting, but I don't think the result would be any better
```cpp
        std::copy_if(active.begin(),active.end(),dwarves,[](df::unit *unit){
            return !Units::isCitizen(unit);
        });
```
It is shorter, but I dunno.. just not as readable imo